### DISCUSSION TITLE
Cannot merge `Router`s whose handlers require application state

### DISCUSSION THREAD

**Discussion by sofiaritz** [tokio-rs/axum#2615] [Q&A] [answered]
### Summary

I'm trying to do something similar to what's done in this example: https://github.com/tokio-rs/axum/blob/main/examples/key-value-store/src/main.rs#L116

But I am getting an error that I cannot fully understand nor fix, I hope someone can shine some light on this issue :)
<details>
<summary>Compilation error</summary>
<pre>
error[E0277]: the trait bound &#x60;Router: From&#x3C;Router&#x3C;AppState&#x3E;&#x3E;&#x60; is not satisfied
   --&#x3E; compmgr_server/src/main.rs:77:16
    |
77  |         .merge(web_index())
    |          ----- ^^^^^^^^^^^ the trait &#x60;From&#x3C;Router&#x3C;AppState&#x3E;&#x3E;&#x60; is not implemented for &#x60;Router&#x60;, which is required by &#x60;Router&#x3C;AppState&#x3E;: Into&#x3C;Router&#x3E;&#x60;
    |          |
    |          required by a bound introduced by this call
    |
    = note: required for &#x60;Router&#x3C;AppState&#x3E;&#x60; to implement &#x60;Into&#x3C;Router&#x3E;&#x60;
note: required by a bound in &#x60;Router::&#x3C;S&#x3E;::merge&#x60;
   --&#x3E; /home/sofia/.cargo/registry/src/index.crates.io-6f17d22bba15001f/axum-0.7.4/src/routing/mod.rs:226:12
    |
224 |     pub fn merge&#x3C;R&#x3E;(self, other: R) -&#x3E; Self
    |            ----- required by a bound in this associated function
225 |     where
226 |         R: Into&#x3C;Router&#x3C;S&#x3E;&#x3E;,
    |            ^^^^^^^^^^^^^^^ required by this bound in &#x60;Router::&#x3C;S&#x3E;::merge&#x60;

error[E0277]: the trait bound &#x60;Router: From&#x3C;Router&#x3C;AppState&#x3E;&#x3E;&#x60; is not satisfied
   --&#x3E; compmgr_server/src/main.rs:78:16
    |
78  |         .merge(other_router(token))
    |          ----- ^^^^^^^^^^^^^^^^^^^^^^^^ the trait &#x60;From&#x3C;Router&#x3C;AppState&#x3E;&#x3E;&#x60; is not implemented for &#x60;Router&#x60;, which is required by &#x60;Router&#x3C;AppState&#x3E;: Into&#x3C;Router&#x3E;&#x60;
    |          |
    |          required by a bound introduced by this call
    |
    = note: required for &#x60;Router&#x3C;AppState&#x3E;&#x60; to implement &#x60;Into&#x3C;Router&#x3E;&#x60;
note: required by a bound in &#x60;Router::&#x3C;S&#x3E;::merge&#x60;
   --&#x3E; /home/sofia/.cargo/registry/src/index.crates.io-6f17d22bba15001f/axum-0.7.4/src/routing/mod.rs:226:12
    |
224 |     pub fn merge&#x3C;R&#x3E;(self, other: R) -&#x3E; Self
    |            ----- required by a bound in this associated function
225 |     where
226 |         R: Into&#x3C;Router&#x3C;S&#x3E;&#x3E;,
    |            ^^^^^^^^^^^^^^^ required by this bound in &#x60;Router::&#x3C;S&#x3E;::merge&#x60;
</pre>
</details>

```rs
// src/main.rs

// A macro to simplify the process of representing `askama` templates
macro_rules! template_route {
    //--snip
    ($name:ident, $path:literal => $file:literal: $struct:ident ($id:literal)) => {
        #[derive(Template)]
        #[template(path = $file)]
        struct $struct<'a> {
            active: &'a str,
        }

        fn $name() -> Router<$crate::AppState> {
            async fn route() -> impl IntoResponse {
                $struct { active: $id }
            }

            Router::<$crate::AppState>::new().route($path, get(route))
        }
    }
}

template_route!(web_index, "/" => "dashboard.html": IndexTemplate ("dashboard"));
template_route!(first_boot_tmpl, "/boot" => "first_boot.html": FirstBootTemplate ("first_boot"));

fn other_router(token: Option<String>) -> Router<crate::AppState> {
    #[debug_handler]
    async fn handler(
        State(state): State<crate::AppState>,
        //--snip
    ) -> Html<&'static str> {
        //--snip
    }

    if token.is_some() {
        Router::<crate::AppState>::new()
            .route("/boot", put(handler))
            .merge(first_boot_tmpl())
    } else {
        Router::<crate::AppState>::new()
    }
}

#[derive(Clone)]
struct AppState {
    // --snip
}

#[tokio::main]
async fn main() {
    // --snip
    let state = AppState {
        // --snip
    };
    let token: Option<String> = // --snip
    let router = Router::new()
        .with_state(state)
        // --snip
        .merge(web_index())
        .merge(other_router(token))
        .layer(TraceLayer::new_for_http())
        .layer(CookieManagerLayer::new());
}
```

### axum version

0.7.4

**Comment by jplatte** [comment, marked as answer]
`with_state` only supplies the state to all routes added prior to the call. You need to supply the state after merging.

**Reply by sofiaritz** [reply]
Edit: Fixed by also explicitly specifying `S` as `AppState` in previous merges, thank you! :)
